ols/utils/config.py

```diff
@@ -4,74 +4,127 @@
+
+    _instance = None  # Class-level variable to hold the singleton instance
+
+    def __new__(cls):
+        """Ensure only one instance of ConfigManager exists (Singleton pattern)."""
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def __init__(self):
+        """Initialize instance attributes only if not already initialized."""
+        if not hasattr(self, "config"):  # Check to avoid re-initialization
+            self.config = None
+        if not hasattr(self, "ols_config"):
+            self.ols_config = None
+        if not hasattr(self, "llm_config"):
+            self.llm_config = None
+        if not hasattr(self, "dev_config"):
+            self.dev_config = None
+        if not hasattr(self, "query_redactor"):
+            self.query_redactor = None
+        if not hasattr(self, "rag_index"):
+            self.rag_index = None
+        if not hasattr(self, "conversation_cache"):
+            self.conversation_cache = None
+
+    def init_empty_config(self) -> None:
+        """Initialize empty configuration."""
+        self.config = config_model.Config()
+        self.ols_config = config_model.OLSConfig()
+        self.llm_config = config_model.LLMProviders()
+        self.dev_config = config_model.DevConfig()
+        print(self.config)
+
+    def init_config(self, config_file: str) -> None:
+        """Load configuration from a YAML file."""
+        try:
+            with open(config_file, "r", encoding="utf-8") as f:
+                self.config = self.load_config_from_stream(f)
+                self.ols_config = self.config.ols_config
+                self.llm_config = self.config.llm_providers
+                self.dev_config = self.config.dev_config
+
+                self.conversation_cache = CacheFactory.conversation_cache(
+                    self.ols_config.conversation_cache
+                )
+        except Exception as e:
+            print(f"Failed to load config file {config_file}: {e!s}")
+            print(traceback.format_exc())
+            raise
+
+    def load_config_from_stream(self, stream: TextIOBase) -> config_model.Config:
+        """Load configuration from a YAML stream."""
+        data = yaml.safe_load(stream)
+        c = config_model.Config(data)
+        c.validate_yaml()
+        return c
+
+    def init_query_filter(self) -> None:
+        """Initialize question filter."""
+        # TODO: OLS-380 Config object mirrors configuration
+        self.query_redactor = QueryFilter()
+
+    def init_vector_index(self) -> None:
+        """Initialize vector index."""
+        # TODO: OLS-380 Config object mirrors configuration
+        self.rag_index = IndexLoader(self.ols_config.reference_content).vector_index
+
+    def get_config(self) -> config_model.Config:
+        """Get the current configuration."""
+        if self.config is None:
+            raise ValueError("Configuration has not been initialized.")
```

Comment:
is not it too late? I mean - we can:

1. have proper config object
2 or the app failed already in `runner.py`

Reply:
i.e. there should not be intermediate state, when the config is half-initialized